upup/pkg/fi/cloudup/apply_cluster.go

```diff
@@ -1399,15 +1398,21 @@ func (n *nodeUpConfigBuilder) BuildConfig(ig *kops.InstanceGroup, apiserverAddit
 	return config, bootConfig, nil
 }
 
-func getTasksCertificate(caTasks map[string]*fitasks.Keypair, name string, config *nodeup.Config, includeKeypairID bool) error {
-	cas, err := fi.ResourceAsString(caTasks[name].Certificates())
+func loadCertificate(keysets map[string]*fi.Keyset, name string, config *nodeup.Config, includeKeypairID bool) error {
+	keyset := keysets[name]
+	if keyset == nil {
+		return fmt.Errorf("key %q not found", name)
+	}
+	certificates, err := keyset.ToCertificateBytes()
 	if err != nil {
-		// CA task may not have run yet; we'll retry
-		return fmt.Errorf("failed to read %s certificates: %w", name, err)
+		return fmt.Errorf("error getting certificate from keyset %q: %w", name, err)
 	}
-	config.CAs[name] = cas
+	config.CAs[name] = string(certificates)
 	if includeKeypairID {
-		config.KeypairIDs[name] = caTasks[name].Keyset().Primary.Id
+		if keyset.Primary == nil {
+			return fmt.Errorf("key %q did not have primary set", name)
```

Comment:
`Keyset` maintains the invariant that there must be a `Primary` if there are any `Keypair`s.

Reply:
Ah - good to know.  I'll revert this and add a comment to Primary (if one is needed!)

Reply:
Actually, as this method returns an error anyway, I think we should keep it - as it's not obvious why the keys in keysets have any keypairs.

Reply:
Keeping it is okay. The keys in keysets have at least one keypair because they come from a `fitasks.Keypair`.

Reply:
Another invariant maintained is that the `Primary` has a non-nil `PrivateKey`.